tests/org.eclipse.jface.tests/src/org/eclipse/jface/tests/viewers/AbstractTreeViewerTest.java

```diff
@@ -164,6 +164,87 @@ public void testExpandToLevel() {
+	}
+
+	@Test
+	public void testSetAutoExpandOnSingleChildLevels() {
+		fTreeViewer.setAutoExpandOnSingleChildLevels(AbstractTreeViewer.NO_EXPAND);
+		assertEquals("Setting an auto-expansion level of NO_EXPAND works",
+				fTreeViewer.getAutoExpandOnSingleChildLevels(), AbstractTreeViewer.NO_EXPAND);
+
+		fTreeViewer.setAutoExpandOnSingleChildLevels(4);
+		assertEquals("Setting a non-trivial auto-expansion level works", fTreeViewer.getAutoExpandOnSingleChildLevels(),
+				4);
+
+		fTreeViewer.setAutoExpandOnSingleChildLevels(AbstractTreeViewer.NO_EXPAND);
+		assertEquals("Setting an auto-expansion level of NO_EXPAND works",
+				fTreeViewer.getAutoExpandOnSingleChildLevels(), AbstractTreeViewer.NO_EXPAND);
+	}
+
+	@Test
+	public void testAutoExpandTrivialPathManualDisable() {
+		// We need our own model since some default models do not generate trivial
+		// paths
+		TestElement modelRoot = TestElement.createModel(5, 1);
+		TestElement trivialPathRoot = modelRoot.getFirstChild();
+		fViewer.setInput(modelRoot);
+
+		fTreeViewer.setAutoExpandOnSingleChildLevels(AbstractTreeViewer.NO_EXPAND);
+		fTreeViewer.setExpandedStateWithAutoExpandOnSingleChild(trivialPathRoot, true);
+
+		assertTrue("The expanded widget child is not expanded", fTreeViewer.getExpandedState(trivialPathRoot));
+		assertTrue("The first child of the trivial path was auto-expanded",
+				fTreeViewer.getExpandedState(trivialPathRoot.getFirstChild()) == false);
+	}
+
+	public void testAutoExpandTrivialPathManualEnableAndThenDisable() {
+		// We need our own model since some default models do not generate trivial
+		// paths
+		TestElement modelRoot = TestElement.createModel(5, 1);
+		TestElement trivialPathRoot = modelRoot.getFirstChild();
+		fViewer.setInput(modelRoot);
+
+		fTreeViewer.setAutoExpandOnSingleChildLevels(2);
+		fTreeViewer.setAutoExpandOnSingleChildLevels(AbstractTreeViewer.NO_EXPAND);
+		fTreeViewer.setExpandedStateWithAutoExpandOnSingleChild(trivialPathRoot, true);
+
+		assertTrue("The expanded widget child is not expanded", fTreeViewer.getExpandedState(trivialPathRoot));
+		assertTrue("The first child of the trivial path was auto-expanded",
+				fTreeViewer.getExpandedState(trivialPathRoot.getFirstChild()) == false);
+	}
+
+	@Test
+	public void testAutoExpandTrivialPathInfiniteExpand() {
+		TestElement modelRoot = TestElement.createModel(5, 1);
+		TestElement trivialPathRoot = modelRoot.getFirstChild();
+		fViewer.setInput(modelRoot);
+
+		fTreeViewer.setAutoExpandOnSingleChildLevels(AbstractTreeViewer.ALL_LEVELS);
+		fTreeViewer.setExpandedStateWithAutoExpandOnSingleChild(trivialPathRoot, true);
+
+		assertTrue("The expanded widget child is not expanded", fTreeViewer.getExpandedState(trivialPathRoot));
+		TestElement child = trivialPathRoot.getFirstChild();
+		int depth = 0;
+		while (child != null) {
+			assertTrue("The " + ++depth + ". child of the trivial path was not auto-expanded",
```

Comment:
To avoid fixed inline incrementation and string generation, you could also do:
```suggestion
for (int depth = 1; child != null; depth++) {
			assertTrue("The " + depth + ". child of the trivial path was not auto-expanded",
```